spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java

```diff
@@ -1252,50 +1276,86 @@ protected Message doSendAndReceiveWithFixed(final String exchange, final String
 		Assert.state(this.isListener, "RabbitTemplate is not configured as MessageListener - "
 							+ "cannot use a 'replyAddress': " + this.replyAddress);
 		return this.execute(channel -> {
-			final PendingReply pendingReply = new PendingReply();
-			String messageTag = String.valueOf(RabbitTemplate.this.messageTagProvider.incrementAndGet());
-			RabbitTemplate.this.replyHolder.put(messageTag, pendingReply);
-			// Save any existing replyTo and correlation data
-			String savedReplyTo = message.getMessageProperties().getReplyTo();
-			pendingReply.setSavedReplyTo(savedReplyTo);
-			if (StringUtils.hasLength(savedReplyTo) && logger.isDebugEnabled()) {
-				logger.debug("Replacing replyTo header:" + savedReplyTo
-						+ " in favor of template's configured reply-queue:"
-						+ RabbitTemplate.this.replyAddress);
-			}
-			message.getMessageProperties().setReplyTo(RabbitTemplate.this.replyAddress);
-			String savedCorrelation = null;
-			if (RabbitTemplate.this.correlationKey == null) { // using standard correlationId property
-				String correlationId = message.getMessageProperties().getCorrelationId();
-				if (correlationId != null) {
-					savedCorrelation = correlationId;
+			return doSendAndReceiveAsListener(exchange, routingKey, message, correlationData, channel);
+		}, obtainTargetConnectionFactory(this.sendConnectionFactorySelectorExpression, message));
+	}
+
+	private Message doSendAndReceiveWithDirect(String exchange, String routingKey, Message message,
+			CorrelationData correlationData) {
+		ConnectionFactory connectionFactory = obtainTargetConnectionFactory(
+				this.sendConnectionFactorySelectorExpression, message);
+		DirectReplyToMessageListenerContainer container = this.directReplyToContainers.get(connectionFactory);
+		if (container == null) {
+			synchronized (this.directReplyToContainers) {
+				if (this.directReplyToContainers.get(connectionFactory) == null) {
+					container = new DirectReplyToMessageListenerContainer(connectionFactory);
+					container.setMessageListener(this);
+					container.start();
+					this.directReplyToContainers.putIfAbsent(connectionFactory, container);
+					this.replyAddress = Address.AMQ_RABBITMQ_REPLY_TO;
 				}
 			}
-			else {
-				savedCorrelation = (String) message.getMessageProperties()
-						.getHeaders().get(RabbitTemplate.this.correlationKey);
-			}
-			pendingReply.setSavedCorrelation(savedCorrelation);
-			if (RabbitTemplate.this.correlationKey == null) { // using standard correlationId property
-				message.getMessageProperties().setCorrelationId(messageTag);
-			}
-			else {
-				message.getMessageProperties().setHeader(RabbitTemplate.this.correlationKey, messageTag);
-			}
+		}
+		ChannelHolder channelHolder = container.getChannelHolder();
```

Comment:
![container_npe](https://cloud.githubusercontent.com/assets/1071590/19566089/0f233526-96b7-11e6-98b5-2b0bf3f69325.png)

And I see how. You don't initialize `container` variable in the `synchronized`.
If `this.directReplyToContainers.get(connectionFactory) != null` we go out from the `synchronized` block, but `container` can be still `null` because that was the reason to step into `synchronized` block.